### PR TITLE
add support for d symbol demangling

### PR DESCRIPTION
**Pull request by lievenhey**

### CODE REVIEW

**Comment by sandsmark** [comment]
Would it make sense to either support piping things through external scripts (a la llvm-symbolizer) or have proper plugins?

It seems awkward manually hardcoding a set of in essence plugins that live half in-tree, half out-of-tree.

It might make sense to have some heuristics for what symbols to pass to the rust or d demanglers, passing every symbol through them seems like unnecessary overhead when 99.9% of the symbols hotspot is going to encounter are neither.

File: app/demangler.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+/* Demangler spezification
+ * int demangler(const char* mangledSymbol, char* demangledBuffer, size_t bufferSize)
+ *
+ * size_t is platform dependent (4 bytes on 32 bit, 8 bytes on 64 bit)
+ * */
+
+namespace {
+bool startsWith(const char* string, const QByteArray& prefix) {
+    if (prefix.isEmpty()) {
```
Comment (milianw) [inline app/demangler.cpp]: use `strcmp(string, prefix.constData()) == 0`;

File: app/demangler.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+Demangler::Demangler()
+{
+    struct DemangleLib {
+        QString name;
+        const char* function;
+        QByteArray prefix;
+    };
+
+    const QVector<DemangleLib> demangleLibs = {
```
Comment (milianw) [inline app/demangler.cpp]: don't use a QVector, just use an initializer list, i.e. just `const auto demangleLibs = {...}`

**Review by milianw — CHANGES_REQUESTED** [review]
generally this looks good to me.

and no, we don't want to spawn external processes, as that's usually quite slow. a library call otoh is fast, and they will have fast-path returns for bad input. we can always optimize this later if need be.

meaning again: I'm perfectly fine with the proposed approach of this patch, it looks good!

**Comment by lievenhey** [comment]
I had completely forgotten this pull request. Fixed those two requests

File: app/demangler.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+namespace {
+bool startsWith(const char* string, const QByteArray& prefix) {
+    return strcmp(string, prefix.constData()) == 0;
+}
+}
+
+Demangler::Demangler()
+{
+    struct DemangleLib {
```
Comment (milianw) [inline app/demangler.cpp]: personally I think it's much simpler to just call the `loadDemangleLib` directly here - then most of this boilerplate can be removed again.

i.e. remove `DemangleLib` and `demangleLibs` and the for loop, and replace it by these two lines:

loadDemangleLib(QStringLiteral("rustc_demangle"), "rustc_demangle", QByteArrayLiteral("_R"));
loadDemangleLib(QStringLiteral("d_demangle"), "demangle_symbol", QByteArrayLiteral("_RD));

File: app/demangler.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+** contact form at http://www.qt.io/contact-us
+**
+****************************************************************************/
+
+#include "demangler.h"
+
+#include <QLibrary>
+#include <QDebug>
+
+/* Demangler spezification
```
Comment (milianw) [inline app/demangler.cpp]: typo: specification

**Review by milianw — CHANGES_REQUESTED** [review]

**Review by milianw — APPROVED** [review]

**Pull request merged**